### PR TITLE
protocol: change the format of handshake packets

### PR DESCRIPTION
**Pull request by aqrln**
* Introduce the concept of authentication strategies.
* Implement `login` and `anonymous` strategies.
* Make handshake packets include the strategy as an action field.
* Update tests.
* Update docs.

Refs: https://github.com/metarhia/JSTP/issues/35#issuecomment-277692315

See more details about the implementation in the updated `doc/protocol.md`.

### CODE REVIEW

File: lib/connection.js [hunk 1]
```diff
@@ -388,26 +390,26 @@ Connection.prototype._processHandshakeRequest = function(packet, keys) {
...
   });
 
-  const callback = this._onSessionCreated.bind(this, username);
-  this.server.startSession(this, application, username, password, callback);
+  this.server.startSession(this, application, authStrategy, credentials,
+    this._onSessionCreated.bind(this));
 };
 
 // Callback of authentication operation
 //   username - user login, if any
```
Comment (belochub) [inline lib/connection.js]: You can probably change order of arguments here as well, to correspond with function signature.

File: lib/simple-auth.js [hunk 2]
```diff
@@ -14,20 +14,23 @@ module.exports = simpleAuth;
 // Start session. Only anonymous handshakes are allowed.
 //   connection - JSTP connection
 //   application - application instance
-//   username - null (required by the auth callback contract)
-//   password - null (required by the auth callback contract)
+//   strategy - authentication strategy (only 'anonymous' is supported)
+//   credentials - authentication credentials
 //   callback - callback function that has signature (error, sessionId)
```
Comment (belochub) [inline lib/simple-auth.js]: Callback signature here does not correspond to the callback you are calling on line 33.

File: lib/simple-auth.js [hunk 3]
```diff
@@ -14,20 +14,23 @@ module.exports = simpleAuth;
...
 //
-simpleAuth.startSession = (connection, application,
-                           username, password, callback) => {
-  if (username) {
+simpleAuth.startSession = (
+  connection, application,
+  strategy, credentials, callback
+) => {
+  console.dir({ strategy, credentials });
```
Comment (belochub) [inline lib/simple-auth.js]: Was it intentionally left here?

Comment (aqrln) [inline lib/simple-auth.js]: Leftover of fixing tests. Thanks for noticing it.

File: doc/protocol.md [hunk 4]
```diff
@@ -129,10 +129,22 @@ Handshake packets always have ID equal to `0`. The response contains either
 the key `ok` with a value that is the session identifier or `error` that is
 an array with error code and optional error message.
 
+The action field of handshake requests specifies the authentication strategy.
+There are two supported strategies now:
+
+* `login` — authentication with login and password. The payload is an array of
+  two elements: user name and password, both represented as strings.
```
Comment (belochub) [inline doc/protocol.md]: In this case it is usually more appropriate to write "username", not "user name".

**Comment by aqrln** [comment]
@belochub updated to resolve the issues and nits.

**Comment by aqrln** [comment]
Rebased to resolve conflicts.

**Comment by aqrln** [comment]
@tshemsedinov awaiting your review.

**Comment by aqrln** [comment]
@tshemsedinov it's semver-major anyway, so it will go along with any other breaking changes. All non-breaking changes are still being backported to v0.6.

**Comment by aqrln** [comment]
@tshemsedinov and it is still possible to do a protocol-level compatibility fixup for v0.6 granted we have no users named 'login', 'anonymous' or 'session' :)
Whom we actually don't even if someone used such nicknames since the protocol-level authentication is not used in the current projects and all users are anonymous from the point of view of JSTP.

**Comment by aqrln** [comment]
@tshemsedinov @belochub it has just occurred to me that in fact the protocol-level compatibility will not be broken for our applications because of the reasons I told in the previous comment. Since only anonymous handshakes were used and authentication was perform using RPC methods, and anonymous handshakes syntax has not changed, we can update the server to JSTP 0.7 without updating the clients.

**Comment by aqrln** [comment]
@tshemsedinov @belochub so basically we don't even need backporting this PR to v0.6 and supporting v0.6 at all.

**Comment by aqrln** [comment]
If we introduce any new breaking changes that will not be feasible to work around, we'll release v0.7 in its current state, call new breaking changes v0.8 and continue landing non-breaking changes to v0.7.

**Comment by aqrln** [comment]
Landed in 9e55d3c11d3be63e49b97485065c9737957f708d.

**Pull request closed without merging**